atomic_reactor/plugins/post_push_to_pulp.py

```diff
@@ -176,11 +179,39 @@ def push_tarball_to_pulp(self, image_names, repo_prefix="redhat-"):
                     tags=[tag]
                 )
 
-        top_layer, layers = self._get_tar_metadata(self.filename)
         self.log.info("pulp_repos = %s", pulp_repos)
         self._create_missing_repos(p, pulp_repos, repo_prefix)
 
-        p.upload(self.filename)
+        top_layer, layers = self._get_tar_metadata(self.filename)
+        # getImageIdsExist was introduced in rh-dockpulp 0.6+
+        try:
+            existing_imageids = p.getImageIdsExist(layers)
+            self.log.debug("existing layers: %s" % existing_imageids)
+
+            # Strip existing layers from the tar and repack it
+            file_path = pipes.quote(self.filename)
+            remove_layers = [pipes.quote(str(os.path.join(x, 'layer.tar'))) for x in existing_imageids]
+
+            p_t = pipes.Template()
+            if file_path.endswith('.xz'):
+                p_t.prepend("xzcat %s" % file_path, ".-")
+            elif file_path.endswith('.gz'):
+                p_t.prepend("zcat %s" % file_path, ".-")
+            elif file_path.endswith('.bz2'):
+                p_t.prepend("bzcat %s" % file_path, ".-")
+            elif file_path.endswith('.tar'):
+                p_t.prepend("cat %s" % file_path, ".-")
+            else:
+                raise Exception("Unknown tarball format: %s" % file_path)
+            p_t.append("tar --delete %s" % ' '.join(remove_layers), "--")
+            p_t.append("gzip -", "--")
+            with p_t.open(NamedTemporaryFile('r', prefix='strip_tar_', delete=True), 'r') as f:
+                f.flush()
+                p.upload(f.name)
+        except Exception:
+            self.log.debug(traceback.format_exc())
```

Comment:
I wonder whether to handle the common case of AttributeError separately to avoid tracebacks in the logs, something like:

```
try:
    get_image_ids = getattr(p, 'getImageIdsExist')
except AttributeError:
    self.log.info("uploading full tar")
    p.upload(self.filename)
else:
    try:
        existing_imageids = get_image_ids(layers)
        ....
    except:
        self.log.debug(traceback.format_exc())
        self.log.info("fallback back to full tar upload")
        p.upload(self.filename)
```